framework/src/media/utils/MediaUtils.cpp

```diff
@@ -43,6 +150,393 @@ static const std::string MP4_MIME_TYPE = "audio/mp4";
+	} // else 1 - 184bytes playload
+	*payloadDataLen = lenPayload;
+	return payload;
+}
+
+// Unpack TS packet to abstract PSI section
+static int PSI_Unpack(const uint8_t *packet, const uint8_t **section, uint16_t *sectlen)
+{
+	uint8_t  lenPayload = 0;
+	const uint8_t *ptrPayload = getPayloadData(packet, &lenPayload);
+	if (!ptrPayload) {
+		// no payload
+		*section = NULL;
+		return UTILS_ERROR_NONE;
+	}
+	if (TS_PAYLOAD_UNIT_START_INDICATOR(packet)) {
+		// new section start, and the 1st byte in payload is the pointer field.
+		uint8_t u8PointerField = ptrPayload[0];
+		if (u8PointerField != 0) {
+			// prev section tail and next section head in this packet!
+			return UTILS_ERROR_NOT_SUPPORTED;
+		}
+		// handle new section
+		const uint8_t *ptrSection = ptrPayload + 1 + u8PointerField;
+		uint16_t lenSection = (SECTION_HEAD_BYTES + SECTION_LENGTH(ptrSection));
+		if (lenSection > (lenPayload - 1 - u8PointerField)) {
+			// portion of section in this packet
+			return UTILS_ERROR_NOT_SUPPORTED;
+		}
+		// TODO: check CRC32
+		// section complete (whole section contained in one TS packet)
+		*section = ptrSection;
+		*sectlen = lenSection;
+		return UTILS_ERROR_NONE;
+	} else {
+		// section appending
+		return UTILS_ERROR_NOT_SUPPORTED;
+	}
+}
+
+// Get the information (number of programs) in the PAT.
+static int ts_psi_get_pat_info(ts_demux_t *demux, psi_pat_info_t *info)
+{
+	psi_section_t *patSect = &demux->pat.section;
+	info->tsId = (patSect->data[3] << 8) + patSect->data[4];
+	// take off 8bytes header and 4bytes CRC, one program consists of 4bytes.
+	info->numOfPrgs = (patSect->size - 12) >> 2;
+	return UTILS_ERROR_NONE;
+}
+
+// Get the program number and PID from PAT at the given program index.
+static int ts_psi_get_program(ts_demux_t *demux, uint16_t index, uint16_t *prgNum, uint16_t *pmtPid)
+{
+	psi_section_t *patSect = &demux->pat.section;
+	if (index >= ((patSect->size - 12) >> 2)) {
+		return UTILS_ERROR_BAD_PARAMETER;
+	}
+	// 8bytes header and 4bytes CRC, loop { 2bytes program number + 2bytes pmt pid }
+	*prgNum = (patSect->data[8 + index * 4] << 8) + patSect->data[9 + index * 4];
+	*pmtPid = ((patSect->data[10 + index * 4] << 8) + patSect->data[11 + index * 4]) & 0x1FFF;
+	return UTILS_ERROR_NONE;
+}
+
+// Get the information (program number, number of streams, etc.) in the PMT.
+static int ts_psi_get_pmt_info(ts_demux_t *demux, psi_pmt_info_t *info)
+{
+	psi_section_t *pmtSect = &demux->pmt.section;
+	info->prgNum = (pmtSect->data[3] << 8) + pmtSect->data[4];
+	info->prgInfoLen = ((pmtSect->data[10] << 8) + pmtSect->data[11]) & 0x0FFF;
+	info->numOfStreams = demux->numOfStreams;
+	return UTILS_ERROR_NONE;
+}
+
+// Get the information (stream type, stream PID, etc.) of the PMT elementary stream.
+static int ts_psi_get_stream_info(ts_demux_t *demux, uint16_t index, psi_pmt_stream_info_t *info)
+{
+	psi_section_t *pmtSect = &demux->pmt.section;
+	if (index >= demux->numOfStreams) {
```

Comment:
there is a IF_FAIL_RETURN_VAL macro defined, but it's not used for these kinds of condition checking in many funcitons. should it be used for them ?

Reply:
Ok, I changed some...
Usually, I use IF_FAIL_RETURN_VAL to check parameters and function return values.